### PR TITLE
feat: enable parsing Bitcoin deposit memo with inscription; deposit fee improvement

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

- [x] Allow parsing Bitcoin deposit memo with inscription. Cherry picked from the original [PR 2727](https://github.com/zeta-chain/node/pull/2727)
- [x] Bitcoin depositor fee improvement to achieve better dev experience. Backport from the original [PR 2765](https://github.com/zeta-chain/node/pull/2765)
- [x] Replaced Solana deprecated RPC `GetRecentBlockhash` with `GetLatestBlockhash` to fix RPC error

# How Has This Been Tested?

- [x] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -1,7 +1,15 @@
 # CHANGELOG
 
+## v19.1.2
```
[the remark this answers is not in the document]
Comment (lumtis) [reply changelog.md:3]: 19.1.1 is the latest released. I think there has been synchorinization issue

File: changelog.md [hunk 2]
```diff
@@ -1,7 +1,15 @@
...
+
+### Features
+
+* [2765](https://github.com/zeta-chain/node/pull/2765) - enable parsing Bitcoin deposit memo with inscription; deposit fee improvement
```
Comment (lumtis) [inline changelog.md:7]:
```suggestion
* [2768](https://github.com/zeta-chain/node/pull/2768) - enable parsing Bitcoin deposit memo with inscription; deposit fee improvement
```

**Review by lumtis — COMMENTED** [review]
Included the E2E test for the inscription please review https://github.com/zeta-chain/node/pull/2769

File: changelog.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (gartnera) [reply changelog.md:3]: I would say 19.2.0 as this is a feature

[the remark this answers is not in the document]
Comment (ws4charlie) [reply changelog.md:3]: @gartnera I can update change log to `v19.2.0`.

File: changelog.md (hunk 2 above)
Comment (ws4charlie) [reply changelog.md]: https://github.com/zeta-chain/node/pull/2768/commits/9a1d22fe05102cd336db11d70c958f94fee79990

File: zetaclient/chains/bitcoin/rpc/rpc.go [hunk 3]
```diff
@@ -63,6 +54,20 @@ func GetTxResultByHash(
 	return hash, txResult, nil
 }
 
+// GetTXRawResultByHash gets the raw transaction by hash
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/rpc/rpc.go]: https://github.com/zeta-chain/node/pull/2768/commits/9a1d22fe05102cd336db11d70c958f94fee79990

File: zetaclient/chains/bitcoin/rpc/rpc.go [hunk 4]
```diff
@@ -118,42 +123,48 @@ func GetRawTxResult(
...
+	// Note: the calculation uses 'Vsize' returned by RPC to simplify dev experience:
+	// 	- 1. the devs could use the same value returned by their RPC endpoints to estimate deposit fee.
+	// 	- 2. the devs don't have to bother 'Vsize' calculation, even though there is more accurate formula.
+	//		 Moreoever, the accurate 'Vsize' is usually an adjusted size (float value) by Bitcoin Core.
+	//	- 3. the 'Vsize' calculation could depend on program language and the library used.
+	//
+	// calculate the fee rate in satoshis/vByte
 	// #nosec G115 always in range
-	return uint64(highestRate), nil
+	feeRate := fee / int64(rawResult.Vsize)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/rpc/rpc.go:167]: Yeah. Tx always has a size. But let me add a ZERO check here even though it should always be positive.

File: zetaclient/chains/bitcoin/observer/witness.go [hunk 5]
```diff
@@ -34,11 +36,26 @@ func GetBtcEventWithWitness(
...
 
-	if err := isValidRecipient(tx.Vout[0].ScriptPubKey.Hex, tssAddress, netParams); err != nil {
+	// check if the recipient is the tss address
+	err := isValidRecipient(tx.Vout[0].ScriptPubKey.Hex, tssAddress, netParams)
+	if err != nil {
 		logger.Debug().Msgf("irrelevant recipient %s for tx %s, err: %s", tx.Vout[0].ScriptPubKey.Hex, tx.Txid, err)
 		return nil, nil
 	}
 
+	// switch to depositor fee V2 if
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/observer/witness.go:46]: We'll remove the version tags once upgrading is done. It's already tracked by https://github.com/zeta-chain/node/issues/2766

File: zetaclient/chains/bitcoin/fee.go [hunk 6]
```diff
@@ -239,3 +253,67 @@ func CalcDepositorFee(
...
+
+	// apply gas price multiplier
+	// #nosec G115 always in range
+	feeRate = int64(float64(feeRate) * clientcommon.BTCOutboundGasPriceMultiplier)
+
+	return DepositorFee(feeRate), nil
+}
+
+// GetRecentFeeRate gets the highest fee rate from recent blocks
+// Note: this method should be used for testnet ONLY
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/fee.go:283]: https://github.com/zeta-chain/node/pull/2768/commits/f23434e6821ce1a34fab1e7de93b43cbf38112d6

File: zetaclient/chains/bitcoin/rpc/rpc.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/rpc/rpc.go:172]: check added: https://github.com/zeta-chain/node/pull/2768/commits/f23434e6821ce1a34fab1e7de93b43cbf38112d6

File: zetaclient/chains/bitcoin/fee.go [hunk 7]
```diff
@@ -239,3 +253,67 @@ func CalcDepositorFee(
 
 	return DepositorFee(feeRate)
 }
+
+// CalcDepositorFeeV2 calculates the depositor fee for a given tx result
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/chains/bitcoin/fee.go:258]: we'll remove the versions once upgrading is done. It was tracked by https://github.com/zeta-chain/node/issues/2766

**Review by lumtis — APPROVED** [review]

**Pull request merged**